File: java/lang/src/main/java/otterop/lang/Result.java
package otterop.lang;

public class Result<RES,ERR> {

    private RES _res;

    private ERR _err;

    private Result(RES _res, ERR _err) {
        this._res = _res;
        this._err = _err;
    }

    public boolean isOK(){
        return  this._err != null;
    }

    public ERR err() {
        return this._err;
    }

    public RES unwrap() {
        return this._res;
    }

    public static <RES0, ERR0> Result<RES0,ERR0> of(RES0 res, ERR0 err) {
        return new Result<RES0,ERR0>(res, err);
    }
}

File: dotnet/src/Lang/Otterop/Lang/Result.cs
namespace Otterop.Lang
{
    public class Result<RES, ERR>
    {
        private RES res;
        private ERR err;
        private Result(RES res, ERR err)
        {
            this.res = res;
            this.err = err;
        }

        public ERR Err()
        {
            return this.err;
        }

        public RES Unwrap()
        {
            return this.res;
        }

        public static Result<RES0, ERR0> Of<RES0, ERR0>(RES0 res, ERR0 err)
        {
            return new Result<RES0, ERR0>(res, err);
        }

    }


    class Result
    {
        public static Result<RES0, ERR0> Of<RES0, ERR0>(RES0 res, ERR0 err)
        {
            return Result<object, object>.Of<RES0, ERR0>(res, err);
        }

    }

}
